src/action.py

```diff
@@ -0,0 +1,344 @@
+
+    if (
+        not draft_mode
+        and action.navlink_change.new.link is not None
+        and action.content_change.new != action.content_change.old
+    ):
+        if action.content_change.new is None:
+            raise exceptions.ActionError(
+                f"internal error, new content for page is None, {action=!r}"
+            )
+
+        try:
+            discourse.update_topic(
+                url=action.navlink_change.new.link, content=action.content_change.new
+            )
+            url = action.navlink_change.new.link
+            result = types_.ActionResult.SUCCESS
+            reason = None
+        except exceptions.DiscourseError as exc:
+            url = action.navlink_change.new.link
+            result = types_.ActionResult.FAIL
+            reason = str(exc)
+    else:
+        if action.navlink_change.new.link is not None:
+            url = DRAFT_NAVLINK_LINK if draft_mode else action.navlink_change.new.link
+        else:
+            # URL for directory should always be None
+            url = None
+        result = types_.ActionResult.SKIP if draft_mode else types_.ActionResult.SUCCESS
+        reason = DRAFT_MODE_REASON if draft_mode else None
+
+    table_row = types_.TableRow(
+        level=action.level, path=action.path, navlink=action.navlink_change.new
+    )
+    return types_.ActionReport(table_row=table_row, url=url, result=result, reason=reason)
+
+
+def _delete(
+    action: types_.DeleteAction, discourse: Discourse, draft_mode: bool, delete_pages: bool
+) -> types_.ActionReport:
+    """Execute a delete action.
+
+    Args:
+        action: The delete action details.
+        discourse: A client to the documentation server.
+        draft_mode: If enabled, only log the action that would be taken.
+        delete_pages: Whether to delete pages that are no longer needed.
+
+    Returns:
+        A report on the outcome of executing the action.
+
+    Raises:
+        ActionError: If the link for a page to delete is None.
+    """
+    logging.info("draft mode: %s, delete pages: %s, action: %s", draft_mode, delete_pages, action)
+
+    is_group = action.navlink.link is None
+    if draft_mode:
+        return types_.ActionReport(
+            table_row=None,
+            url=action.navlink.link,
+            result=types_.ActionResult.SKIP,
+            reason=DRAFT_MODE_REASON,
+        )
+    if not delete_pages and not is_group:
+        return types_.ActionReport(
+            table_row=None,
+            url=action.navlink.link,
+            result=types_.ActionResult.SKIP,
+            reason=NOT_DELETE_REASON,
+        )
+    if is_group:
+        return types_.ActionReport(
+            table_row=None,
+            url=action.navlink.link,
+            result=types_.ActionResult.SUCCESS,
+            reason=None,
+        )
+
+    try:
+        # Edge case that should not be possible
+        if action.navlink.link is None:  # pragma: no cover
```

Comment:
The reason this if statement is here is for defensive programming, to make sure that if some mistake occurs because, for example, `is_group` was changed in some way, an error is raised. The call to `discourse.delete_topic` depends on `action.navlink.link` not being `None`, rather than passing `None` to that function and having to figure out why `None` was passed, this if statement ensures a nice error is raised to show what has gone wrong. It is related to the philosophy if raising errors as early as possible

Reply:
Ok, that's what I guessed, but wanted to make sure :)